### DISCUSSION TITLE
Is this library similar to Tone.js?

### DISCUSSION THREAD

**Discussion by lumosminima** [elemaudio/elementary#45] [Q&A] [answered]
Hi 👋, newbie here. I've previously used [Tone.js](https://tonejs.github.io/) and it's [`Tone.Sampler`](https://tonejs.github.io/docs/14.7.77/Sampler.html) to create basic midi playable online piano. I was wondering where does this library fit is it lower level than tone.js? From some research it seems like this doesn't use much of web audio API and has it's own C++ engine which I would assume improves performance and is awesome. I was looking for a more performant version of something like Tone.js which has lots of helpers etc understanding of 'Notes' and triggering like `sampler.triggerAttackRelease('C2')` but not sure if much of this would need to be implemented by developer or the library provides some of this. It would also be helpful to add a section in docs comparing `elementary` to other libraries newbies think of when "web" and "audio" are the topic i.e `Tone.js`,`Howler.js` etc

**Comment by nick-thompson** [comment, marked as answer]
Hey @lumosminima thank you for the question and I'm terribly sorry for the delay. This library is indeed similar to Tone, and I think you've correctly pointed out some of the differences.

> I was wondering where does this library fit is it lower level than tone.js? From some research it seems like this doesn't use much of web audio API and has it's own C++ engine which I would assume improves performance and is awesome.

It is lower level in that sense, yes! Tone is written using many of the web audio API nodes, whereas Elementary uses only a single WebAudioWorklet (when running in the web browser), and that worklet is running the C++ engine compiled to WASM. Because Elementary targets that C++ engine, we can also run the same code on other platforms (i.e. compile it in a native desktop app, audio plugin, raspberry pi, etc).

> something like Tone.js which has lots of helpers etc understanding of 'Notes' and triggering like sampler.triggerAttackRelease('C2') but not sure if much of this would need to be implemented by developer or the library provides some of this.

This piece is very much something I want Elementary to have soon (and is on the roadmap!), but it doesn't exactly exist yet. You can absolutely write a sampler and a synthesizer with Elementary right now, but the javascript library is currentlymissing some of these higher level utilities that make working with synths/samplers/notes/sequences as easy as Tone.js. So my short answer there is stay tuned, but also feel free to dig into the lower level stuff if you're interested!

> It would also be helpful to add a section in docs comparing elementary to other libraries newbies think of when "web" and "audio" are the topic i.e Tone.js,Howler.js etc

This is a great idea, I've just added a note in my todo list to write this up :+1:
